### PR TITLE
Refactor parent map

### PR DESCRIPTION
**Pull request by sbihel**
Address #364

### CODE REVIEW

**Comment by danglotb** [comment]
Hi @sbihel 

Thank you again for the PR. It seems that we still have the same issues.

After investigation, the problem comes from the fact that we modify the CtMethod used as a key.

We clone the given CtMethod to modify it. Since we put the key just after clone, and before modification, the `HashMap` computes the hashcode of the unmodified clone and so when we want to get using the modified clone, no hashcode match, and we obtain a NPE.

A simple solution is to use an IdentityHashMap rather than a HashMap.

I just opened a PR on your branch, see sbihel/dspot#1

I am gonna review the rest of your PR after the merging.

**Comment by sbihel** [comment]
Aaah I see. Silly me

Thanks a lot!

File: dspot/src/main/java/fr/inria/diversify/utils/AmplificationHelper.java [hunk 1]
```diff
@@ -290,7 +319,8 @@ public static String getClassPath(DSpotCompiler compiler, InputConfiguration con
             final Long average = average(valuesToMethod.keySet());
             while (reducedTests.size() < MAX_NUMBER_OF_TESTS) {
                 final Long furthest = furthest(valuesToMethod.keySet(), average);
-                reducedTests.add(valuesToMethod.get(furthest).get(0));
+                CtMethod<?> discardedTest = valuesToMethod.get(furthest).get(0);
+                reducedTests.add(discardedTest);
```
Comment (danglotb) [inline dspot/src/main/java/fr/inria/diversify/utils/AmplificationHelper.java]: The name is wrongly chosen. In the opposite that happen: we keep the "furthest" method. That is why I removed your call to `remove()` on the map of parents. I think we can inline the variable here.

Comment (sbihel) [reply dspot/src/main/java/fr/inria/diversify/utils/AmplificationHelper.java]: I have to be more focused. Sorry

Comment (danglotb) [reply dspot/src/main/java/fr/inria/diversify/utils/AmplificationHelper.java]: No you good, That's the whole point of the review process :)

**Pull request merged**